bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptDependencyTracker.java

```diff
@@ -0,0 +1,138 @@
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.automation.module.script.ScriptDependencyTracker;
+import org.openhab.core.automation.module.script.rulesupport.internal.loader.collection.BidiSetBag;
+import org.openhab.core.service.AbstractWatchService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link AbstractScriptDependencyTracker} tracks dependencies between scripts and reloads dependees
+ * It needs to be sub-classed for each {@link org.openhab.core.automation.module.script.ScriptEngineFactory}
+ * that wants to support dependency tracking
+ *
+ * @author Jonathan Gilbert - Initial contribution
+ * @author Jan N. Klug - Refactored to OSGi service
+ */
+@NonNullByDefault
+public abstract class AbstractScriptDependencyTracker implements ScriptDependencyTracker {
+    private final Logger logger = LoggerFactory.getLogger(AbstractScriptDependencyTracker.class);
+
+    protected final String libraryPath;
+
+    private final Set<ScriptDependencyTracker.Listener> dependencyChangeListeners = ConcurrentHashMap.newKeySet();
+
+    private final BidiSetBag<String, String> scriptToLibs = new BidiSetBag<>();
+    private @Nullable AbstractWatchService dependencyWatchService;
+
+    public AbstractScriptDependencyTracker(final String libraryPath) {
+        this.libraryPath = libraryPath;
+    }
+
+    public void activate() {
+        AbstractWatchService dependencyWatchService = createDependencyWatchService();
+        dependencyWatchService.activate();
+        this.dependencyWatchService = dependencyWatchService;
+    }
+
+    public void deactivate() {
+        AbstractWatchService dependencyWatchService = this.dependencyWatchService;
+        if (dependencyWatchService != null) {
+            dependencyWatchService.deactivate();
+        }
+    }
+
+    protected AbstractWatchService createDependencyWatchService() {
+        return new AbstractWatchService(libraryPath) {
+            @Override
+            protected boolean watchSubDirectories() {
+                return true;
+            }
+
+            @Override
+            protected WatchEvent.Kind<?> @Nullable [] getWatchEventKinds(Path path) {
+                return new WatchEvent.Kind<?>[] { ENTRY_CREATE, ENTRY_DELETE, ENTRY_MODIFY };
+            }
+
+            @Override
+            protected void processWatchEvent(WatchEvent<?> watchEvent, WatchEvent.Kind<?> kind, Path path) {
+                File file = path.toFile();
+                if (!file.isHidden() && (kind.equals(ENTRY_DELETE)
+                        || (file.canRead() && (kind.equals(ENTRY_CREATE) || kind.equals(ENTRY_MODIFY))))) {
+                    AbstractScriptDependencyTracker.this.dependencyChanged(file.getPath());
```

Comment:
Would this also work:

```suggestion
                    dependencyChanged(file.getPath());
```